allennlp/interpret/attackers/input_reduction.py

```diff
@@ -0,0 +1,144 @@
+            current_tokens = current_instances[0][name_of_input_field_to_attack].tokens
+            smallest_idx = -1
+            # keep removing tokens
+            while len(current_instances[0][name_of_input_field_to_attack]) >= num_ignore_tokens:
+                # get gradients and predictions
+                grads, outputs = self.predictor.get_gradients(current_instances)
+                for output in outputs:
+                    if isinstance(outputs[output], torch.Tensor):
+                        outputs[output] = outputs[output].detach().cpu().numpy().squeeze().squeeze()
+                    elif isinstance(outputs[output], list):
+                        outputs[output] = outputs[output][0]
+
+                # Check if any fields have changed, if so, break loop
+                if "tags" not in current_instances[0]:
+                    equal = True
+                    for field in fields_to_check:
+                        if field in current_instances[0].fields:
+                            equal = equal and current_instances[0][field].__eq__(fields_to_check[field])
+                        else:
+                            equal = equal and outputs[field] == fields_to_check[field]
+                    if not equal:
+                        break
+
+                # special case for sentence tagging (we have tested NER)
+                else:
+                    if smallest_idx != -1:
+                        del tag_mask[smallest_idx]
+                    cur_tags = [outputs["tags"][x] for x in range(len(outputs["tags"])) if tag_mask[x]]
+                    if cur_tags != original_tags:
+                        break
+
+                # remove a token from the input
+                current_tokens = list(current_instances[0][name_of_input_field_to_attack].tokens)
+                current_instances, smallest_idx = \
+                    remove_one_token(grads[name_of_grad_input_field],
+                                     current_instances,
+                                     name_of_input_field_to_attack, ignore_tokens)
+
+            final_tokens.append(current_tokens)
+        return sanitize({"final": final_tokens, "original": original_tokens})
+
+def remove_one_token(grads: np.ndarray,
+                     instances: List[Instance] = None,
+                     name_of_input_field_to_attack: str = 'tokens',
+                     ignore_tokens: List[str] = ["@@NULL@@"]) -> List[Instance]: # pylint: disable=dangerous-default-value
+    """
+    Finds the token with the smallest gradient and removes it.
+    """
+    # Compute L2 norm of all grads.
+    grads_mag = [np.sqrt(grad.dot(grad)) for grad in grads]
+
+    # Skip all ignore_tokens by setting grad to infinity
+    for tok_idx, tok in enumerate(instances[0][name_of_input_field_to_attack].tokens):
+        if tok in ignore_tokens:
+            grads_mag[tok_idx] = float("inf")
+    # For NER, skip all tokens that are not in outside
```

Comment:
It's unfortunate that we have to special case NER throughout here.  I'm not sure there's a way around it, other than having `predictions_to_labeled_instances` have a more complex return type, that tells you stuff about which tokens shouldn't be changed.